src/java.base/unix/classes/java/lang/ProcessEnvironment.java

```diff
@@ -62,27 +68,92 @@
 {
     private static final HashMap<Variable,Value> theEnvironment;
     private static final Map<String,String> theUnmodifiableEnvironment;
+    // CRIU capable
+    private static final boolean criuCapable;
+    // 1 - prints a message if the env var was set but not in the immutable list
+    // 2 - throws an exception
+    private static final int tracePrunedEnvVarsValue;
+    private static final Map<String,String> theOriginalUnmodifiableEnvironment;
     static final int MIN_NAME_LENGTH = 0;
 
     static {
+        criuCapable = "true".equalsIgnoreCase(System.internalGetProperties().getProperty("org.eclipse.openj9.criu.isCRIUCapable"));
+        HashMap<Variable,Value> criuEnvironment = null;
+        Set<String> criuImmutableEnvVarList = null;
+        if (criuCapable) {
+            String strTracePrunedEnvVars = System.internalGetProperties().getProperty("org.eclipse.openj9.criu.TracePrunedEnvVars");
+            if (strTracePrunedEnvVars != null) {
+                tracePrunedEnvVarsValue = Integer.valueOf(strTracePrunedEnvVars);
+                if (tracePrunedEnvVarsValue > 2 || tracePrunedEnvVarsValue < 1) {
+                    throw new InternalError("CRIU: org.eclipse.openj9.criu.TracePrunedEnvVars unexpected value: " + String.valueOf(tracePrunedEnvVarsValue));
+                }
+            } else {
+                tracePrunedEnvVarsValue = 0;
+            }
+            // CRIU immutable env var list
+            String propImmutableEnvVars = System.internalGetProperties().getProperty("org.eclipse.openj9.criu.ImmutableEnvVars");
+            String[] immutableEnvVarArray = (propImmutableEnvVars == null) ? new String[] {} : propImmutableEnvVars.split(",");
+            criuImmutableEnvVarList = new HashSet<String>(Arrays.asList(immutableEnvVarArray));
+            // hardcoded list
+            criuImmutableEnvVarList.add("LANG");
+            criuImmutableEnvVarList.add("LC_ALL");
+            criuImmutableEnvVarList.add("LC_CTYPE");
+            criuEnvironment = new HashMap<>(criuImmutableEnvVarList.size() + 3);
+        } else {
+            tracePrunedEnvVarsValue = 0;
+        }
         // We cache the C environment.  This means that subsequent calls
         // to putenv/setenv from C will not be visible from Java code.
         byte[][] environ = environ();
         theEnvironment = new HashMap<>(environ.length/2 + 3);
         // Read environment variables back to front,
         // so that earlier variables override later ones.
+        Variable tmpKeyVar;
         for (int i = environ.length-1; i > 0; i-=2)
-            theEnvironment.put(Variable.valueOf(environ[i-1]),
-                               Value.valueOf(environ[i]));
+        {
+            tmpKeyVar = Variable.valueOf(environ[i-1]);
+            if (criuCapable) {
+                for (String envVar : criuImmutableEnvVarList) {
```

Comment:
since its a `Set` we dont need to interate, we can use `contains()`

Reply:
yeah, switched to `contains()`.